client/transaction-pool/src/lib.rs

```diff
@@ -727,6 +806,128 @@ where
 	}
 }
 
+/// Helper struct for deciding if core part of maintenance procedure shall be executed.
+/// It is publicly exposed only for tests purposes (due to cyclic deps sc_transaction_pool and
+/// substrate_test_runtime_transaction_pool).
+///
+/// For the following chain:
+///   B1-C1-D1-E1
+///  /
+/// A
+///  \
+///   B2-C2-D2-E2
+///
+/// Some scenarios and expected behavior:
+/// nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(C1))
+/// f(C1), nbb(C1) -> false (handle_enactment was already performed in f(C1))
+///
+/// f(C1), nbb(D2) -> false (handle_enactment was already performed in f(C1), we should not retract
+/// finalized block)
+/// f(C1), f(C2), nbb(C1) -> false
+/// nbb(C1), nbb(C2) -> true (switching fork is OK)
+/// nbb(B1), nbb(B2) -> true
+/// nbb(B1), nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(B1)
+/// nbb(C1), f(B1) -> false (handle_enactment was already performed in nbb(B2)
+#[doc(hidden)]
+pub struct EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	recent_best_block: Option<Block::Hash>,
+	recent_finalized_block: Option<Block::Hash>,
+}
+
+impl<Block> EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	fn new() -> Self {
+		EnactmentState { recent_best_block: None, recent_finalized_block: None }
+	}
+
+	pub fn update_and_check_if_new_enactment_is_valid<PoolApi: graph::ChainApi<Block = Block>>(
+		&mut self,
+		api: &PoolApi,
+		event: &ChainEvent<Block>,
+	) -> Result<(bool, Option<TreeRoute<Block>>), String> {
+		let (new_hash, finalized) = match event {
+			ChainEvent::NewBestBlock { hash, .. } => (*hash, false),
+			ChainEvent::Finalized { hash, .. } => (*hash, true),
+		};
+
+		let best_block = match self.recent_best_block {
+			Some(recent_best_hash) => recent_best_hash,
+			None => {
+				self.recent_best_block = Some(new_hash);
+				if finalized {
+					self.recent_finalized_block = Some(new_hash);
+				}
+
+				return Ok((true, None))
+			},
+		};
+
+		// compute actual tree route from best_block to notified block, and use it instead of
+		// tree_route provided with event
+		let tree_route = match api.tree_route(best_block, new_hash) {
+            Ok(tree_route) => {
+                tree_route .expect("tree_route exists. qed.")
```

Comment:
Is it just github or is the formatting fucked up?

Reply:
This `expect` is also really bad here. As said already, we should not return an optional tree route.

Reply:
`cargo +nightly fmt --all` did not fix that.

Reply:
[cfd7f63](https://github.com/paritytech/substrate/pull/12305/commits/cfd7f63953f45079e1a0bbab08718755b6188364)